File: apps/server/src/routes/export.ts
import { Router, Request, Response, NextFunction } from 'express';
import { requireAuth } from '../middleware/auth';
import { createValidationError, createInternalError, createNotFoundError } from '../middleware/error-handler';
import { logger } from '../utils/logger';
import { PlexClient, resolvePlexToken } from '../services/plex';
import { updateNotification } from '../services/job-notifications';
import { enqueueAction } from '../services/action-queue';

const router = Router();

// All export routes require authentication
router.use(requireAuth);

// originalTitle is the real per-track artist; grandparentTitle is just the
// album/folder artist and can be wrong for compilations/soundtracks.
function getTrackArtist(track: any): string {
  return track.originalTitle || track.grandparentTitle || 'Unknown Artist';
}

// Export is two Plex reads + in-memory string building, not a Plex write or
// external API call - its cost doesn't scale with concurrent users the way
// the queue's other actions do, so only genuinely large exports (which take
// long enough to be worth not blocking on) go through the action queue.
// Below this, export stays fully synchronous - same request/response as
// always, no client changes.
const EXPORT_QUEUE_THRESHOLD = 2000;

interface ExportedFile {
  filename: string;
  contentType: string;
  content: string;
}

function buildExportContent(tracks: any[], playlistTitle: string, format: string, pathType: string): ExportedFile {
  let content: string;
  let contentType: string;
  let extension: string;

  switch (format) {
    case 'm3u':
      content = generateM3U(tracks, playlistTitle, pathType === 'relative');
      contentType = 'audio/x-mpegurl';
      extension = 'm3u';
      break;
    case 'm3u8':
      content = generateM3U8(tracks, playlistTitle, pathType === 'relative');
      contentType = 'application/vnd.apple.mpegurl';
      extension = 'm3u8';
      break;
    case 'pls':
      content = generatePLS(tracks, playlistTitle);
      contentType = 'audio/x-scpls';
      extension = 'pls';
      break;
    case 'xspf':
      content = generateXSPF(tracks, playlistTitle);
      contentType = 'application/xspf+xml';
      extension = 'xspf';
      break;
    case 'csv':
      content = generateCSV(tracks);
      contentType = 'text/csv';
      extension = 'csv';
      break;
    case 'txt':
      content = generateTXT(tracks);
      contentType = 'text/plain';
      extension = 'txt';
      break;
    default:
      throw new Error('Unsupported format');
  }

  return { filename: `${sanitizeFilename(playlistTitle)}.${extension}`, contentType, content };
}

// Finished exports queued above the threshold, held just long enough for the
// client to fetch them via GET /export/:jobId/download - same in-memory,
// lost-on-restart tradeoff as the rest of the action queue.
const pendingExports = new Map<string, ExportedFile>();
const EXPORT_RESULT_TTL_MS = 10 * 60 * 1000;

/**
 * POST /api/playlists/export
 * Export a playlist to various formats
 */
router.post('/export', async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { playlistId, format, pathType } = req.body;

    if (!playlistId || typeof playlistId !== 'string') {
      return next(createValidationError('playlistId is required'));
    }

    if (!format || !['m3u', 'm3u8', 'pls', 'xspf', 'csv', 'txt'].includes(format)) {
      return next(createValidationError('Invalid format'));
    }

    const userId = req.session.userId!;
    const db = req.dbService!;

    // Get user's Plex token
    const userRow = (db as any).db.prepare('SELECT plex_token FROM users WHERE id = ?').get(userId);
    if (!userRow) {
      return next(createValidationError('User not found'));
    }

    const { plex_token: accountToken } = userRow;
    if (!accountToken) {
      return next(createValidationError('No Plex token found'));
    }

    // Get user's server configuration
    const serverRow = (db as any).db.prepare('SELECT server_url, access_token FROM user_servers WHERE user_id = ? LIMIT 1').get(userId);
    if (!serverRow) {
      return next(createValidationError('No Plex server configured'));
    }

    const { server_url: serverUrl } = serverRow;

    // Create Plex client
    const plexClient = new PlexClient(serverUrl, resolvePlexToken({ plex_token: accountToken }, serverRow));

    // Get playlist details
    const playlistDetails = await plexClient.getPlaylistDetails(playlistId);
    if (!playlistDetails) {
      return next(createValidationError('Playlist not found'));
    }

    if ((playlistDetails.leafCount || 0) > EXPORT_QUEUE_THRESHOLD) {
      const { jobId, position } = enqueueAction(userId, `Export: ${playlistDetails.title}`, async (notificationId) => {
        try {
          const tracks = await plexClient.getPlaylistTracks(playlistId);
          const file = buildExportContent(tracks, playlistDetails.title, format, pathType);
          pendingExports.set(notificationId, file);
          setTimeout(() => pendingExports.delete(notificationId), EXPORT_RESULT_TTL_MS);
          logger.info(`Exported playlist: ${playlistDetails.title} (${tracks.length} tracks) to ${format}`);
          updateNotification(userId, notificationId, { status: 'success', detail: `Ready to download - ${file.filename}` });
        } catch (error: any) {
          logger.error('Failed to export playlist', { error: error.message, userId });
          updateNotification(userId, notificationId, { status: 'error', detail: error.message || 'Failed to export playlist' });
        }
      });
      res.status(202).json({ queued: true, jobId, position });
      return;
    }

    // Get playlist tracks
    const tracks = await plexClient.getPlaylistTracks(playlistId);

    logger.info(`Exporting playlist: ${playlistDetails.title} (${tracks.length} tracks) to ${format}`);

    const { filename, contentType, content } = buildExportContent(tracks, playlistDetails.title, format, pathType);

    // Send the file
    res.setHeader('Content-Type', contentType);
    res.setHeader('Content-Disposition', `attachment; filename="${filename}"`);
    res.send(content);

  } catch (error: any) {
    logger.error('Failed to export playlist', { error: error.message });
    next(createInternalError(`Failed to export playlist: ${error.message}`));
  }
});

/**
 * GET /api/playlists/export/:jobId/download
 * Fetches the file generated by a queued export (see EXPORT_QUEUE_THRESHOLD
 * above) once its notification reports success. `jobId` is that
 * notification's id, scoped to no one but the requester by virtue of only
 * ever being handed back to the user who queued the export in the first
 * place - it's never exposed to other users.
 */
router.get('/export/:jobId/download', (req: Request, res: Response, next: NextFunction) => {
  const file = pendingExports.get((req.params as Record<string, string>).jobId);
  if (!file) {
    return next(createNotFoundError('Export not found or already downloaded - it may have expired, try exporting again'));
  }
  pendingExports.delete((req.params as Record<string, string>).jobId);
  res.setHeader('Content-Type', file.contentType);
  res.setHeader('Content-Disposition', `attachment; filename="${file.filename}"`);
  res.send(file.content);
});

/**
 * Generate M3U format
 */
function generateM3U(tracks: any[], playlistName: string, relative: boolean): string {
  let content = '#EXTM3U\n';
  content += `#PLAYLIST:${playlistName}\n`;

  for (const track of tracks) {
    const duration = Math.floor((track.duration || 0) / 1000);
    const artist = getTrackArtist(track);
    const title = track.title || 'Unknown Track';
    const filePath = getTrackPath(track, relative);

    content += `#EXTINF:${duration},${artist} - ${title}\n`;
    content += `${filePath}\n`;
  }

  return content;
}

/**
 * Generate M3U8 format (UTF-8 encoded M3U)
 */
function generateM3U8(tracks: any[], playlistName: string, relative: boolean): string {
  // M3U8 is the same as M3U but with UTF-8 encoding
  return generateM3U(tracks, playlistName, relative);
}

/**
 * Generate PLS format
 */
function generatePLS(tracks: any[], playlistName: string): string {
  let content = '[playlist]\n';
  content += `PlaylistName=${playlistName}\n`;
  content += `NumberOfEntries=${tracks.length}\n\n`;

  tracks.forEach((track, index) => {
    const num = index + 1;
    const filePath = getTrackPath(track, false);
    const title = `${getTrackArtist(track)} - ${track.title || 'Unknown'}`;
    const duration = Math.floor((track.duration || 0) / 1000);

    content += `File${num}=${filePath}\n`;
    content += `Title${num}=${title}\n`;
    content += `Length${num}=${duration}\n\n`;
  });

  content += 'Version=2\n';
  return content;
}

/**
 * Generate XSPF format (XML)
 */
function generateXSPF(tracks: any[], playlistName: string): string {
  let content = '<?xml version="1.0" encoding="UTF-8"?>\n';
  content += '<playlist version="1" xmlns="http://xspf.org/ns/0/">\n';
  content += `  <title>${escapeXml(playlistName)}</title>\n`;
  content += '  <trackList>\n';

  for (const track of tracks) {
    const filePath = getTrackPath(track, false);
    const title = track.title || 'Unknown Track';
    const artist = getTrackArtist(track);
    const album = track.parentTitle || 'Unknown Album';
    const duration = track.duration || 0;

    content += '    <track>\n';
    content += `      <location>file://${escapeXml(filePath)}</location>\n`;
    content += `      <title>${escapeXml(title)}</title>\n`;
    content += `      <creator>${escapeXml(artist)}</creator>\n`;
    content += `      <album>${escapeXml(album)}</album>\n`;
    content += `      <duration>${duration}</duration>\n`;
    if (track.index) {
      content += `      <trackNum>${track.index}</trackNum>\n`;
    }
    content += '    </track>\n';
  }

  content += '  </trackList>\n';
  content += '</playlist>\n';
  return content;
}

/**
 * Generate CSV format
 */
function generateCSV(tracks: any[]): string {
  let content = 'Track,Artist,Album,Duration,File Path\n';

  for (const track of tracks) {
    const title = escapeCSV(track.title || 'Unknown Track');
    const artist = escapeCSV(getTrackArtist(track));
    const album = escapeCSV(track.parentTitle || 'Unknown Album');
    const duration = formatDuration(track.duration || 0);
    const filePath = escapeCSV(getTrackPath(track, false));

    content += `${title},${artist},${album},${duration},${filePath}\n`;
  }

  return content;
}

/**
 * Generate plain-text format: one "Artist - Title" per line, matching what
 * this app's own .txt file importer (scrapers.ts's parseM3UFile fallback
 * path) expects, so a playlist exported to .txt can be re-imported unchanged.
 */
function generateTXT(tracks: any[]): string {
  return tracks
    .map(track => `${getTrackArtist(track)} - ${track.title || 'Unknown Track'}`)
    .join('\n') + '\n';
}

/**
 * Get the file path for a track
 */
function getTrackPath(track: any, relative: boolean): string {
  const part = track.Media?.[0]?.Part?.[0];
  if (!part || !part.file) {
    return '';
  }

  let filePath = part.file;

  if (relative) {
    // Convert to relative path (remove leading slash and add ../)
    filePath = filePath.replace(/^\//, '../');
  }

  return filePath;
}

/**
 * Sanitize filename for safe file system usage
 */
function sanitizeFilename(filename: string): string {
  return filename
    .replace(/[<>:"/\\|?*]/g, '_')
    .replace(/\s+/g, '_')
    .substring(0, 200);
}

/**
 * Escape XML special characters
 */
function escapeXml(str: string): string {
  return str
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;');
}

/**
 * Escape CSV special characters
 */
function escapeCSV(str: string): string {
  if (str.includes(',') || str.includes('"') || str.includes('\n')) {
    return `"${str.replace(/"/g, '""')}"`;
  }
  return str;
}

/**
 * Format duration in milliseconds to MM:SS
 */
function formatDuration(ms: number): string {
  const seconds = Math.floor(ms / 1000);
  const minutes = Math.floor(seconds / 60);
  const remainingSeconds = seconds % 60;
  return `${minutes}:${remainingSeconds.toString().padStart(2, '0')}`;
}

export default router;

File: internal/handlers/export.go
// Playlist export, ported from routes/export.ts: write a playlist's tracks
// out as a downloadable file in one of six formats.
//
// Not ported: the Node route's EXPORT_QUEUE_THRESHOLD (playlists over 2000
// tracks built the file in the background via the action queue, then served
// it from a separate GET .../download endpoint once ready). That threshold
// existed to keep a large in-memory string build from blocking Node's
// single-threaded event loop for every other request. Go handlers run one
// per goroutine, so a slow export - the GetPlaylistTracks call to Plex, not
// the string building itself, is the only part that takes real time -
// blocks only its own request, not the server. The pendingExports map/TTL
// and the extra route this bought in Node bought nothing here, so this
// route always replies with the finished file directly, however large.
package handlers

import (
	"bytes"
	"encoding/csv"
	"fmt"
	"net/http"
	"regexp"
	"strings"

	"github.com/go-chi/chi/v5"

	"github.com/drevilish/playlist-lab/internal/auth"
	"github.com/drevilish/playlist-lab/internal/services/plex"
)

func RegisterExport(r chi.Router, mw *auth.Middleware, h *PlaylistsHandler) {
	r.Group(func(r chi.Router) {
		r.Use(mw.RequireAuth)
		r.Post("/playlists/{plexId}/export", h.export)
	})
}

var exportFormats = map[string]string{
	"m3u":  "audio/x-mpegurl",
	"m3u8": "application/vnd.apple.mpegurl",
	"pls":  "audio/x-scpls",
	"xspf": "application/xspf+xml",
	"csv":  "text/csv",
	"txt":  "text/plain",
}

func (h *PlaylistsHandler) export(w http.ResponseWriter, r *http.Request) {
	user := auth.CurrentUser(r)
	plexID := chi.URLParam(r, "plexId")
	_ = r.ParseForm()
	format := r.FormValue("format")
	contentType, ok := exportFormats[format]
	if !ok {
		names := make([]string, 0, len(exportFormats))
		for k := range exportFormats {
			names = append(names, k)
		}
		http.Error(w, "format must be one of: "+strings.Join(names, ", "), http.StatusBadRequest)
		return
	}
	// "relative" only affects M3U/M3U8 (generateM3U's own `relative` flag in
	// the original); every other format always wrote absolute paths, ported
	// as-is below.
	relative := r.FormValue("pathType") == "relative"

	client, userServer, err := h.client(user)
	if err != nil || userServer == nil {
		http.Error(w, "no server selected", http.StatusBadRequest)
		return
	}

	name := plexID
	if playlists, err := client.GetPlaylists(); err == nil {
		for _, p := range playlists {
			if p.RatingKey == plexID {
				name = p.Title
				break
			}
		}
	}

	tracks, err := client.GetPlaylistTracks(plexID)
	if err != nil {
		http.Error(w, "Failed to load playlist tracks", http.StatusBadGateway)
		return
	}

	content := buildExportContent(tracks, name, format, relative)
	filename := sanitizeFilename(name) + "." + format
	w.Header().Set("Content-Type", contentType)
	w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="%s"`, filename))
	w.Write([]byte(content))
}

func buildExportContent(tracks []plex.Track, playlistName, format string, relative bool) string {
	switch format {
	case "m3u", "m3u8":
		return generateM3U(tracks, playlistName, relative)
	case "pls":
		return generatePLS(tracks, playlistName)
	case "xspf":
		return generateXSPF(tracks, playlistName)
	case "csv":
		return generateCSV(tracks)
	case "txt":
		return generateTXT(tracks)
	default:
		return ""
	}
}

// exportArtist prefers the track-level artist (OriginalTitle) over the
// album/grandparent artist, ports export.ts's getTrackArtist - deliberately
// its own function rather than reusing Track.DisplayArtist(), which also
// treats an OriginalTitle equal to GrandparentTitle as "no better answer";
// export.ts's getTrackArtist has no such equality check, so this matches it
// exactly rather than introducing a behavior difference in exported files.
func exportArtist(t plex.Track) string {
	if t.OriginalTitle != "" {
		return t.OriginalTitle
	}
	if t.GrandparentTitle != "" {
		return t.GrandparentTitle
	}
	return "Unknown Artist"
}

func exportTitle(t plex.Track) string {
	if t.Title != "" {
		return t.Title
	}
	return "Unknown Track"
}

// exportPath is export.ts's getTrackPath: relative rewrites a leading "/"
// to "../" (the on-disk path as Plex sees it, relative to Plex's own
// library root, going up one level to wherever the exported file is
// expected to sit alongside a copy of that library layout) - a blunt
// transform ported unchanged from the original rather than reworked, since
// there's no way to know a given user's actual playback-device layout to
// do better.
func exportPath(t plex.Track, relative bool) string {
	path := t.FilePath()
	if path == "" {
		return ""
	}
	if relative {
		return "../" + strings.TrimPrefix(path, "/")
	}
	return path
}

func generateM3U(tracks []plex.Track, playlistName string, relative bool) string {
	var b strings.Builder
	b.WriteString("#EXTM3U\n")
	fmt.Fprintf(&b, "#PLAYLIST:%s\n", playlistName)
	for _, t := range tracks {
		seconds := t.Duration / 1000
		fmt.Fprintf(&b, "#EXTINF:%d,%s - %s\n", seconds, exportArtist(t), exportTitle(t))
		fmt.Fprintf(&b, "%s\n", exportPath(t, relative))
	}
	return b.String()
}

func generatePLS(tracks []plex.Track, playlistName string) string {
	var b strings.Builder
	b.WriteString("[playlist]\n")
	fmt.Fprintf(&b, "PlaylistName=%s\n", playlistName)
	fmt.Fprintf(&b, "NumberOfEntries=%d\n\n", len(tracks))
	for i, t := range tracks {
		num := i + 1
		// export.ts's PLS title falls back to plain "Unknown" rather than
		// "Unknown Track" (every other format's fallback, via exportTitle) -
		// kept as its own inline fallback to match that inconsistency
		// exactly rather than silently "fixing" it into a different output.
		title := t.Title
		if title == "" {
			title = "Unknown"
		}
		fmt.Fprintf(&b, "File%d=%s\n", num, exportPath(t, false))
		fmt.Fprintf(&b, "Title%d=%s - %s\n", num, exportArtist(t), title)
		fmt.Fprintf(&b, "Length%d=%d\n\n", num, t.Duration/1000)
	}
	b.WriteString("Version=2\n")
	return b.String()
}

func generateXSPF(tracks []plex.Track, playlistName string) string {
	var b strings.Builder
	b.WriteString(`<?xml version="1.0" encoding="UTF-8"?>` + "\n")
	b.WriteString(`<playlist version="1" xmlns="http://xspf.org/ns/0/">` + "\n")
	fmt.Fprintf(&b, "  <title>%s</title>\n", xspfEscape(playlistName))
	b.WriteString("  <trackList>\n")
	for _, t := range tracks {
		album := t.ParentTitle
		if album == "" {
			album = "Unknown Album"
		}
		b.WriteString("    <track>\n")
		fmt.Fprintf(&b, "      <location>file://%s</location>\n", xspfEscape(exportPath(t, false)))
		fmt.Fprintf(&b, "      <title>%s</title>\n", xspfEscape(exportTitle(t)))
		fmt.Fprintf(&b, "      <creator>%s</creator>\n", xspfEscape(exportArtist(t)))
		fmt.Fprintf(&b, "      <album>%s</album>\n", xspfEscape(album))
		fmt.Fprintf(&b, "      <duration>%d</duration>\n", t.Duration)
		if t.Index != 0 {
			fmt.Fprintf(&b, "      <trackNum>%d</trackNum>\n", t.Index)
		}
		b.WriteString("    </track>\n")
	}
	b.WriteString("  </trackList>\n")
	b.WriteString("</playlist>\n")
	return b.String()
}

func xspfEscape(s string) string {
	r := strings.NewReplacer(
		"&", "&amp;",
		"<", "&lt;",
		">", "&gt;",
		`"`, "&quot;",
		"'", "&apos;",
	)
	return r.Replace(s)
}

// generateCSV uses encoding/csv rather than porting export.ts's hand-rolled
// escapeCSV: the stdlib writer already handles quoting/escaping correctly
// per RFC 4180, including the embedded-newline case escapeCSV's simple
// "contains a comma, quote, or newline" check also caught, so there's no
// reason to hand-write it again.
func generateCSV(tracks []plex.Track) string {
	var buf bytes.Buffer
	writer := csv.NewWriter(&buf)
	writer.Write([]string{"Track", "Artist", "Album", "Duration", "File Path"})
	for _, t := range tracks {
		album := t.ParentTitle
		if album == "" {
			album = "Unknown Album"
		}
		writer.Write([]string{
			exportTitle(t), exportArtist(t), album,
			formatDuration(t.Duration), exportPath(t, false),
		})
	}
	writer.Flush()
	return buf.String()
}

func formatDuration(ms int64) string {
	seconds := ms / 1000
	return fmt.Sprintf("%d:%02d", seconds/60, seconds%60)
}

// generateTXT matches this app's own .txt importer (fileimport.go's plain-
// text fallback path) so a playlist exported to .txt round-trips through
// re-import unchanged.
func generateTXT(tracks []plex.Track) string {
	var b strings.Builder
	for _, t := range tracks {
		fmt.Fprintf(&b, "%s - %s\n", exportArtist(t), exportTitle(t))
	}
	return b.String()
}

var unsafeFilenameChars = regexp.MustCompile(`[<>:"/\\|?*]`)
var whitespaceRun = regexp.MustCompile(`\s+`)

func sanitizeFilename(name string) string {
	name = unsafeFilenameChars.ReplaceAllString(name, "_")
	name = whitespaceRun.ReplaceAllString(name, "_")
	if len(name) > 200 {
		name = name[:200]
	}
	return name
}
